contracts/feirari/AutoRewardsDistributor.sol

```diff
@@ -0,0 +1,102 @@
+        bool _isBorrowIncentivized
+    ) CoreRef(coreAddress) {
+        isBorrowIncentivized = _isBorrowIncentivized;
+        cTokenAddress = _cTokenAddress;
+        tribalChiefRewardIndex = _tribalChiefRewardIndex;
+        rewardsDistributorAdmin = _rewardsDistributorAdmin;
+        tribalChief = _tribalChief;
+    }
+
+    function _deriveRequiredCompSpeed() internal view returns (uint256 compSpeed) {
+        /// @notice call the TribalChief and grab the current allocation points for the stakedTokenWrapper
+        (,,, uint120 allocPoints,) = tribalChief.poolInfo(tribalChiefRewardIndex);
+        /// @notice get the total allocation points in the TribalChief
+        uint256 totalAllocPoints = tribalChief.totalAllocPoint();
+        /// @notice get the amount of tribe being distributed on a per block basis in the TribalChief
+        uint256 tribePerBlock = tribalChief.tribePerBlock();
+
+        /// @notice if tribe per block, total allocation points or this pools allocation points are equal to 0,
+        /// then set compSpeed to 0
+        if (totalAllocPoints == 0) {
+            compSpeed = 0;
+        } else {
+            /// @notice calculate the amount per block that the Rari rewards distributor is receiving from the TribalChief
+            compSpeed = (tribePerBlock * allocPoints) / totalAllocPoints;
+        }
+    }
+
+    /// @notice function to get the new comp speed and figure out if an update is needed
+    function getRewardSpeedDifference() public view returns (uint256 newCompSpeed, bool updateNeeded) {
+        newCompSpeed = _deriveRequiredCompSpeed();
+        uint256 actualCompSpeed;
+
+        if (isBorrowIncentivized) {
+            actualCompSpeed = rewardsDistributorAdmin.compBorrowSpeeds(cTokenAddress);
+        } else {
+            actualCompSpeed = rewardsDistributorAdmin.compSupplySpeeds(cTokenAddress);
+        }
+
+        /// if actual comp speed is not equal to required comp speed, then an update is needed
+        if (actualCompSpeed != newCompSpeed) {
+            updateNeeded = true;
+        }
+    }
+
+    /// @notice function to automatically set the rewards speed on the RewardsDistributor contract
+    /// through the RewardsDistributorAdmin
+    function setAutoRewardsDistribution() external whenNotPaused {
+        (uint256 compSpeed, bool updateNeeded) = getRewardSpeedDifference();
+        require(updateNeeded, "AutoRewardsDistributor: update not needed");
+
+        /// @notice call out to the rewards distributor admin and set the comp supply/borrow speed to the current value
+        /// if the contract is borrow incentivized, set that speed, else, set the supply speed
+        if (isBorrowIncentivized) {
+            rewardsDistributorAdmin._setCompBorrowSpeed(cTokenAddress, compSpeed);
+        } else {
+            rewardsDistributorAdmin._setCompSupplySpeed(cTokenAddress, compSpeed);
+        }
+        emit SpeedChanged(isBorrowIncentivized, compSpeed, cTokenAddress);
```

Comment:
isborrowIncentivized is immutable state, same with cTokenAddress. These can be removed from the event. Normally events only want to emit changed/mutable state

Reply:
done